test_project/test_app/tests/tests_recipe_serializers.py

```diff
@@ -0,0 +1,133 @@
+            {
+                'ingredients': [ErrorDetail(string='Ensure that any recipe at least has one ingredient defined.', code='invalid')]
+            }
+        )
+
+    def test_recipe_deserializer_with_invalid_ingredients_for_a_drink_recipe(self):
+        """
+        Test recipe deserializer with invalid ingredients for a drink recipe.
+        The test will ensure that any object-level validation inside `clean()` method will be triggered
+        when `serializer.is_valid()` performed.
+        """
+        payload = {
+            "name": "Raspberry Orange",
+            "type": Recipe.DRINK,
+            "ingredients": "This only contain one ingredient"
+        }
+
+        serializer = RecipeSerializer(data=payload)
+
+        with self.assertRaises(ValidationError) as ctx:
+            serializer.is_valid(raise_exception=True)
+
+        self.assertDictEqual(
+            ctx.exception.detail,
+            {
+                'ingredients': [ErrorDetail(string='Any drink recipe at least has two ingredients are defined.', code='invalid')]
+            }
+        )
+
+    def test_recipe_deserializer_with_invalid_ingredients_for_a_main_dish_recipe(self):
+        """
+        Test recipe deserializer with invalid ingredients for a main dish recipe.
+        The test will ensure that any object-level validation inside `clean()` method will be triggered
+        when `serializer.is_valid()` performed.
+        """
+        payload = {
+            "name": "Baked Teriyaki Chicken",
+            "type": Recipe.MAIN_DISH,
+            "ingredients": "First ingredient#Second ingredient"
+        }
+
+        serializer = RecipeSerializer(data=payload)
+
+        with self.assertRaises(ValidationError) as ctx:
+            serializer.is_valid(raise_exception=True)
+
+        self.assertDictEqual(
+            ctx.exception.detail,
+            {
+                'ingredients': [ErrorDetail(string='Any main dish recipe at least has three ingredients are defined.', code='invalid')]
+            }
+        )
+
+    def test_recipe_deserializer_with_valid_payload(self):
+        """Test recipe deserializer with valid payload, data successfully saved into database"""
+        ingredients = ('1 tablespoon cornstarch#'
+                       '1 tablespoon cold water#'
+                       '1/2 cup white sugar#'
+                       '1/2 cup soy sauce#'
+                       '1/4 cup cider vinegar#'
+                       '1 clove garlic, minced#'
+                       '1/2 teaspoon ground ginger#'
+                       '1/4 teaspoon ground black pepper#'
+                       '12 skinless chicken thighs')
+        payload = {
+            "name": "Baked Teriyaki Chicken",
+            "type": Recipe.MAIN_DISH,
+            "ingredients": ingredients
+        }
+
+        serializer = RecipeSerializer(data=payload)
+        serializer.is_valid(True)
+        serializer.save()
+
+        recipe = Recipe.objects.first()
```

Comment:
You don't need to save the instance and then retrieve it. You can simply check `serializer.validated_data`.